pkg/adaptor/mongodb/bulk.go

```diff
@@ -0,0 +1,127 @@
+package mongodb
+
+import (
+	"sync"
+	"time"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/ops"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var (
+	_ client.Writer = &Bulk{}
+)
+
+// Bulk implements client.Writer for use with MongoDB and takes advantage of the Bulk API for
+// performance improvements.
+type Bulk struct {
+	bulkMap   map[string]*bulkOperation
+	lock      sync.Mutex
+	opCounter int
+}
+
+type bulkOperation struct {
+	c          *mgo.Collection
+	insertDocs []interface{}
+	updateDocs []interface{}
+	deleteDocs []interface{}
+}
+
+type writeMsg struct {
+	msg message.Msg
+	s   *Session
+}
+
+func newBulker(done chan struct{}, wg *sync.WaitGroup) *Bulk {
+	b := &Bulk{
+		bulkMap: make(map[string]*bulkOperation),
+	}
+	go b.run(done, wg)
+	return b
+}
+
+func (b *Bulk) Write(msg message.Msg) func(client.Session) error {
+	return func(s client.Session) error {
+		db, coll, _ := message.SplitNamespace(msg)
+		b.lock.Lock()
+		bOp, ok := b.bulkMap[coll]
+		if !ok {
+			bOp = &bulkOperation{
+				c:          s.(*Session).mgoSession.Copy().DB(db).C(coll),
+				insertDocs: make([]interface{}, 0),
+				updateDocs: make([]interface{}, 0),
+				deleteDocs: make([]interface{}, 0),
+			}
+			b.bulkMap[coll] = bOp
+		}
+		b.lock.Unlock()
+		switch msg.OP() {
+		case ops.Delete:
+			bOp.deleteDocs = append(bOp.deleteDocs, bson.M{"_id": msg.Data().Get("_id")})
+		case ops.Insert:
+			bOp.insertDocs = append(bOp.insertDocs, msg.Data())
+		case ops.Update:
+			bOp.updateDocs = append(
+				bOp.updateDocs,
+				bson.M{"_id": msg.Data().Get("_id")},
+				msg.Data(),
+			)
+		}
+		b.opCounter++
+		if b.opCounter == 1000 {
+			b.flush()
+		}
+		return nil
+	}
+}
+
+func (b *Bulk) run(done chan struct{}, wg *sync.WaitGroup) {
+	wg.Add(1)
+	for {
+		select {
+		case <-time.After(2 * time.Second):
+			b.flush()
+		case <-done:
+			log.Debugln("received done channel")
+			b.flush()
+			wg.Done()
+			return
+		}
+	}
+}
+
+func (b *Bulk) flush() error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	log.Debugln("flushing bulk messages")
+	for coll, bulkOp := range b.bulkMap {
+		bulk := bulkOp.c.Bulk()
+		if len(bulkOp.insertDocs) > 0 {
```

Comment:
dang... yes, very good catch and I need to adjust the tests to mix insert/update/deletes ops